### DISCUSSION TITLE
Keystone with pm2 or forever

### DISCUSSION THREAD

**Discussion by shivathejas** [keystonejs/keystone#7560] [Questions] [answered]
How to use Keystone V5 with pm2 or forever to make application up and running forever in production?

**Comment by justintemps** [comment]
The simplest way is simply to have an [ecosystem.config.js](https://pm2.keymetrics.io/docs/usage/application-declaration/) file like this:

```
module.exports = {
  apps: [
    {
      name: "keystone-project",
      cwd: "/path/to/keystone/project/",
      script: "npm run build && npm start",
      instances: 1,
      autorestart: true,
      max_memory_restart: "1G",
      env: {
        DATABASE_URL: "mongodb://localhost/keystone-project-db"
      }
    }
  ]
};
```

Then you'd simply start it with `pm2 start ecosystem.config.js`

This works fine for me, with one caveat: I have no successfully been able to run KeystoneJS in cluster mode.

**Comment by MohammadKurjieh** [comment, marked as answer]
I was able to get v6 running in a pm2 cluster, not sure if this applies to v5. But I am leaving this here in case someone stumbles upon this in the future.

My ecosystem.config.js:

```
module.exports = {
  apps: [
    {
      name: 'my-keystone',
      script: './node_modules/@keystone-6/core/bin/cli.js',
      args: 'start',

      instances: 1,
      exec_mode: 'cluster',
      wait_ready: true,
      listen_timeout: 10000,

      env: {
        PORT: 3000,
      },
    },
  ],
};
```

Note that you need to run build before reloading in pm2.

We are using clustering in pm2 to achieve 0 downtime reloads.

Build: `npm run build` | `yarn build`
Start: `pm2 start ecosystem.config.js`
Reload: `pm2 reload "my-keystone" --update-env`
